src/object.hpp

```diff
@@ -498,6 +524,229 @@ class object_view final : public readable_object<object_view> {
+};
+
+static_assert(sizeof(array_view) <= 16);
+static_assert(sizeof(array_view::iterator) <= 16);
+
+class map_view {
+public:
+    map_view() = default;
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    map_view(const detail::object *o)
+    {
+        if (o == nullptr || o->type != object_type::map) {
+            throw std::invalid_argument("map_view initialised with null or incompatible type");
+        }
+
+        data_ = o->via.map.ptr;
+        size_ = o->via.map.size;
+    }
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    map_view(const detail::object &o) : map_view(&o) {}
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    map_view(object_view o) : map_view(o.ptr()) {}
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    map_view(const owned_object &ow);
+    // NOLINTNEXTLINE(google-explicit-constructor, hicpp-explicit-conversions)
+    map_view(const borrowed_object &ow);
+
+    map_view(owned_object &&ow) = delete;
+
+    ~map_view() = default;
+    map_view(const map_view &) = default;
+    map_view(map_view &&) = default;
+    map_view &operator=(const map_view &) = default;
+    map_view &operator=(map_view &&) = default;
+
+    [[nodiscard]] std::size_t size() const noexcept
+    {
+        // NOLINTNEXTLINE(clang-analyzer-core.uninitialized.UndefReturn)
+        return static_cast<std::size_t>(size_);
+    }
+
+    [[nodiscard]] bool empty() const noexcept { return size() == 0; }
+
+    // Access the value at index.
+    [[nodiscard]] std::pair<object_view, object_view> at(std::size_t index) const noexcept
+    {
+        assert(index < size() && data_ != nullptr);
+        const auto &kv = data_[index];
+        return {kv.key, kv.val};
+    }
+
+    [[nodiscard]] object_view at_value(std::size_t index) const noexcept
+    {
+        assert(index < size() && data_ != nullptr);
+        return data_[index].val;
+    }
+
+    [[nodiscard]] object_view at_key(std::size_t index) const noexcept
+    {
+        assert(index < size() && data_ != nullptr);
+        return data_[index].key;
+    }
+
+    [[nodiscard]] object_view find(std::string_view expected_key) const noexcept
+    {
+        for (std::size_t i = 0; i < size(); ++i) {
+            auto [key, value] = at(i);
+
+            if (expected_key == key.as<std::string_view>()) {
+                return value;
+            }
+        }
+        return {};
+    }
+
+    object_view find_key_path(std::span<const std::string> key_path)
+    {
+        auto root = *this;
+        for (auto it = key_path.begin(); it != key_path.end(); ++it) {
+            object_view child = root.find(*it);
+            if ((it + 1) == key_path.end()) {
+                return child;
+            }
+
+            if (!child.has_value() || !child.is_map()) {
+                break;
+            }
+
+            root = child;
+        }
+        return {};
+    }
```

Comment:
I think this is both too complex an operation to have here and at the same time could be more useful if it allowed also array traversal.

Reply:
Maybe, I don't think I'm using it at the moment. I need to support globs and such so I might remove it and write something more elaborate.